test/communication/thallium_protocol_test.sh

```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+SERVER_BIN="/mnt/common/kfeng/CLionProjects/ChronoLog_dev/decouple_conf_manager/ChronoLog/build/test/communication/thallium_server"
+CLIENT_BIN="/mnt/common/kfeng/CLionProjects/ChronoLog_dev/decouple_conf_manager/ChronoLog/build/test/communication/thallium_client"
+SERVER_HOST="ares-comp-03"
+CLIENT_HOST="ares-comp-04"
+SERVER_IP=$(ssh ${SERVER_HOST} "ip addr show | grep 172.25 | awk '{print $2}' | sed 's/^[[:space:]]*//' | cut -d' ' -f2 | cut -d'/' -f1") # only works for Ares
+SERVER_PORT=5555
+MODES=("sendrecv" "rdma")
+PROTOS=("sockets" "tcp" "verbs")
+SIZE=134217728
+REP=1
+
+for proto in "${PROTOS[@]}"
+do
+  for mode in "${MODES[@]}"
+  do
+    addr=ofi+${proto}://${SERVER_IP}:${SERVER_PORT}
+    echo server address: ${addr}
+    echo -e "\033[0;33mLaunching server ...\033[0m"
+    ssh ${SERVER_HOST} killall -9 thallium_server > /dev/null 2>&1
+    ssh ${SERVER_HOST} "ABT_THREAD_STACKSIZE=2097152 ${SERVER_BIN} \"${addr}\" 1 1" &
+    sleep 3
+    pid=$!
+    echo -e "\033[0;33mLaunching client ...\033[0m"
+    ssh ${CLIENT_HOST} killall -9 thallium_client > /dev/null 2>&1
+    timeout -s 9 10 ssh ${CLIENT_HOST} ${CLIENT_BIN} \"${addr}\" ${mode} ${SIZE} ${REP}
+    ret=$?
+    kill -9 ${pid}
+    if [[ ${ret} -eq 0 ]]
+    then
+      echo -e "\033[0;32mTest of procotol ${proto} in mode ${mode} PASSES\033[0m"
```

Comment:
Instead of setting values manually, can you use slurm commands and executable names to make this script general? It should be able to run on any number of nodes on ares. You can use one node as the server and the others as clients.

You can use this deployment script for other tests on Ares, by replacing the nested loop.

Reply:
You don't have to sbatch the script. Since slurm is installed on the machine, you can run slurm queries from scripts after salloc succeeds. Use slurm commands to query hostnames instead of setting them manually. scontrol show job jobid will have the nodelist. You can parse the output of this command to fill an array of hostnames. Choose one of them as server and the others as clients.  Port number is program specific. You can choose how to pass that. 

Your script can take 3-4 arguments, i.e server_bin, client_bin, number of clients, jobid. 
./script.sh server_bin, client_bin, n, jobid. 

Would have been nicer with a scripting language. I am not sure if you want to do this in this issue or create a separate issue for deployment scripts.